### PR TITLE
Add betainc to JAX

### PR DESCRIPTION
**Pull request by srvasude**
Add regularized incomplete beta to JAX.

### CODE REVIEW

**Comment by srvasude** [comment]
Hrmm it seems that Travis fails (it can't find RegularizedIncompleteBeta), but running this locally passes.

**Comment by hawkinsp** [comment]
I think the issue is that we haven't made a binary release of `jaxlib` since incomplete beta functions were added. Travis tests with the up to date Python `jax` but with a prebuilt `jaxlib`. We're probably due for another `jaxlib` release pretty soon.

The easiest workaround for now is to skip the test if the jaxlib version is <= 0.1.37 (the current release), as is done here: https://github.com/google/jax/blob/11224bd2b19a011cd65577940ac26eb80be04ba7/tests/linalg_test.py#L77

File: jax/lax/lax.py [hunk 1]
```diff
@@ -192,6 +192,14 @@ def atan2(x, y):
     :math:`\mathrm{atan}({x \over y})`."""
   return atan2_p.bind(x, y)
 
+def betainc(a, b, x):
+  r"""Elementwise regularized incomplete beta integral."""
+  # TODO(srvasude): Remove these once XLA broadcasts these inputs.
+  a = _brcast(_brcast(a, b), x)
```
Comment (hawkinsp) [inline jax/lax/lax.py]: As a general rule, `lax` primitives don't have broadcasting semantics themselves (except for scalars). There are a few exceptions. We tend to try to expose XLA fairly directly. It would be more common to support broadcasting in the `jax.scipy.special` wrapper, if that's what scipy itself does.

File: jax/interpreters/masking.py [hunk 2]
```diff
@@ -348,6 +348,29 @@ def binop_masking_rule(prim, padded_vals, logical_shapes):
   padded_x, padded_y = padded_vals
   return prim.bind(padded_x, padded_y)
 
+def defterop(prim):
```
Comment (hawkinsp) [inline jax/interpreters/masking.py]: Could we generalize this to an nary rule rather than having two similar rules?

**Comment by srvasude** [comment]
Thanks! Just added a guard in lax_scipy_test.py.

File: jax/lax/lax.py [hunk 3]
```diff
@@ -1603,6 +1611,16 @@ def binop(result_dtype, accepted_dtypes, name, translation_rule=None):
   return prim
 standard_binop = partial(binop, _input_dtype)
 
+def terop(result_dtype, accepted_dtypes, name, translation_rule=None):
```
Comment (hawkinsp) [inline jax/lax/lax.py]: If you merge the ternary and binary cases in `masking.py`, these two rules are then identical and can be merged, I think?

File: jax/lax/lax.py [hunk 4]
```diff
@@ -1701,6 +1719,9 @@ def _sign_translation_rule(c, x):
   lambda g, x, y: _brcast(g, y) * (y / (square(x) + square(y))),
   lambda g, x, y: _brcast(g, x) * -x / (square(x) + square(y)))
 
+regularized_incomplete_beta_p = standard_terop(
+    [_float, _float, _float], 'regularized_incomplete_beta')
+
```
Comment (hawkinsp) [inline jax/lax/lax.py]: It looks like it wouldn't be hard to define a JVP rule for betainc too? TF's gradient appears to be in terms of `lgamma`, which is available here.

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: jax/lax/lax.py (hunk 4 above)
Comment (srvasude) [reply jax/lax/lax.py]: Is there a way to mark it for one variable? The gradient for x is defined, but the gradient for a and b require a custom special function

Comment (hawkinsp) [reply jax/lax/lax.py]: The closest would be to raise an exception if the unhandled tangent inputs are anything other than the object `ad_util.zero`.

Comment (srvasude) [reply jax/lax/lax.py]: Added it for x, and raised an exception. Is this what you had in mind?

File: jax/lax/lax.py [hunk 5]
```diff
@@ -192,6 +192,11 @@ def atan2(x, y):
     :math:`\mathrm{atan}({x \over y})`."""
   return atan2_p.bind(x, y)
 
+def betainc(a, b, x):
+  r"""Elementwise regularized incomplete beta integral."""
+  # TODO(srvasude): Remove these once XLA broadcasts these inputs.
```
Comment (hawkinsp) [inline jax/lax/lax.py]: Remove comment?

File: jax/lax/lax.py [hunk 6]
```diff
@@ -1696,11 +1701,31 @@ def _sign_translation_rule(c, x):
...
+def betainc_gradx(g, a, b, x):
+  lbeta = lgamma(a) + lgamma(b) - lgamma(a + b)
+  partial_x = exp((b - 1) * log1p(-x) +
+                  (a - 1) * log(x) - lbeta)
+  g = _brcast(_brcast(g, a), b)
+  return partial_x * g
+
+def raise_if_tangent_not_zero(g, a, b, x):
+  if g is ad_util.zero:
+    return _brcast(_brcast(_brcast(g, a), b), x)
```
Comment (hawkinsp) [inline jax/lax/lax.py]: This looks wrong (surely the output of a zero pertubation is zero also.) But it looks like the `defjvp` logic already handles symbolic zeros for you. So you can simplify this to just raising an exception. If this function is called, you are in the case you don't handle.

File: jax/lax/lax.py [hunk 7]
```diff
@@ -1696,11 +1701,31 @@ def _sign_translation_rule(c, x):
...
   lambda g, x, y: _brcast(g, x) * -x / (square(x) + square(y)))
 
+regularized_incomplete_beta_p = standard_naryop(
+    [_float, _float, _float], 'regularized_incomplete_beta')
+
+def betainc_gradx(g, a, b, x):
+  lbeta = lgamma(a) + lgamma(b) - lgamma(a + b)
+  partial_x = exp((b - 1) * log1p(-x) +
+                  (a - 1) * log(x) - lbeta)
+  g = _brcast(_brcast(g, a), b)
```
Comment (hawkinsp) [inline jax/lax/lax.py]: Is this necessary? The primal operation didn't support broadcasting so why would the gradient need to?

File: jax/scipy/special.py [hunk 8]
```diff
@@ -39,6 +39,15 @@ def betaln(x, y):
   return lax.lgamma(x) + lax.lgamma(y) - lax.lgamma(x + y)
 
 
+@_wraps(osp_special.betainc)
+def betainc(a, b, x):
+  a, b, x = _promote_args_inexact("betainc", a, b, x)
+  a = lax._brcast(lax._brcast(a, b), x)  # pylint: disable=protected-access
```
Comment (hawkinsp) [inline jax/scipy/special.py]: I actually want to take this back. You should move this broadcasting back to `lax`. `lax` primitives do usually support broadcasting, just not rank promotion.

File: tests/lax_test.py [hunk 9]
```diff
@@ -114,6 +114,8 @@ def op_record(op, nargs, dtypes, rng_factory, tol=None):
     op_record("sinh", 1, float_dtypes + complex_dtypes, jtu.rand_default),
     op_record("cosh", 1, float_dtypes + complex_dtypes, jtu.rand_default),
 
+    op_record("betainc", 3, float_dtypes, jtu.rand_positive,
```
Comment (hawkinsp) [inline tests/lax_test.py]: Probably need to guard this with a version test too.

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/lax/lax.py (hunk 7 above)
Comment (hawkinsp) [reply jax/lax/lax.py]: Note even when `lax` op did support broadcasting, it was done in the wrapper function, not in the primitive's implementation. So the primitive itself didn't broadcast and nor will its gradient need to.

File: jax/lax/lax.py [hunk 10]
```diff
@@ -192,6 +192,13 @@ def atan2(x, y):
     :math:`\mathrm{atan}({x \over y})`."""
   return atan2_p.bind(x, y)
 
+def betainc(a, b, x):
+  r"""Elementwise regularized incomplete beta integral."""
+  a = _brcast(lax._brcast(a, b), x)
```
Comment (hawkinsp) [inline jax/lax/lax.py]: One final nit: drop the `lax.` here because we're in the `lax` module.

**Review by hawkinsp — APPROVED** [review]
